### PR TITLE
impl shadow realm as js builtins

### PR DESCRIPTION
**Pull request by philomates**

### CODE REVIEW

File: Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+void ShadowRealmPrototype::finishCreation(VM& vm)
+{
+    Base::finishCreation(vm);
+    ASSERT(inherits(vm, info()));
+    JSC_TO_STRING_TAG_WITHOUT_TRANSITION();
+}
+
+JSC_DEFINE_HOST_FUNCTION(importInRealm, (JSGlobalObject* globalObject, CallFrame* callFrame))
```
Comment (philomates) [inline Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:62]: mostly copied and adapted from https://github.com/WebKit/WebKit/blob/fca70ae1e2cf0eeb3532d29a16e0f22a273ee3c5/Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp#L813-L835

File: Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    JSValue thisValue = callFrame->uncheckedArgument(0);
+    ShadowRealmObject* thisRealm = jsDynamicCast<ShadowRealmObject*>(vm, thisValue);
+    if (UNLIKELY(!thisRealm))
+        return throwVMTypeError(globalObject, scope, "'this' should be a ShadowRealm");
+
+    auto* promise = JSPromise::create(vm, globalObject->promiseStructure());
+
+    auto sourceOrigin = callFrame->callerSourceOrigin(vm);
+    auto* specifier = callFrame->uncheckedArgument(1).toString(globalObject);
+    RETURN_IF_EXCEPTION(scope, { });
```
Comment (philomates) [inline Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:81]: note the difference from https://github.com/WebKit/WebKit/blob/main/Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp#L824
In the case here, according to the test262 suite, we want to throw outside the promise if `toString` has issues. I think the difference here achieves that (at least I had to make the change to get things passing)

File: Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+    if (callFrame->argumentCount() != 2) {
+        throwTypeError(globalObject, scope, "Expected to be called with two arguments"_s);
+        RELEASE_AND_RETURN(scope, JSValue::encode(jsUndefined()));
+    }
+
+    JSValue thisValue = callFrame->uncheckedArgument(0);
+    ShadowRealmObject* thisRealm = jsDynamicCast<ShadowRealmObject*>(vm, thisValue);
+    if (UNLIKELY(!thisRealm))
+        return throwVMTypeError(globalObject, scope, "'this' should be a ShadowRealm");
```
Comment (philomates) [inline Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:75]: when should one use `throwVMTypeError` instead of `throwTypeError`?

File: Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+    auto* promise = JSPromise::create(vm, globalObject->promiseStructure());
+
+    auto sourceOrigin = callFrame->callerSourceOrigin(vm);
+    auto* specifier = callFrame->uncheckedArgument(1).toString(globalObject);
+    RETURN_IF_EXCEPTION(scope, { });
+
+    JSGlobalObject* realmGlobalObject = thisRealm->globalObject();
+    auto* internalPromise = realmGlobalObject->moduleLoader()->importModule(realmGlobalObject, specifier, jsUndefined(), sourceOrigin);
```
Comment (philomates) [inline Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:84]: this creates an internal promise somehow associated with the realm global object, but we later do `resolve` on the original global object. I can't see anything clearly problematic with this, but felt worth bringing attention to this

File: Source/JavaScriptCore/runtime/JSGlobalObject.cpp [hunk 5]
```diff
@@ -2550,6 +2570,13 @@ JSGlobalObject* JSGlobalObject::create(VM& vm, Structure* structure)
     return globalObject;
 }
 
+JSGlobalObject* JSGlobalObject::create(VM& vm, Structure* structure, const GlobalObjectMethodTable* methodTable)
+{
+    JSGlobalObject* globalObject = new (NotNull, allocateCell<JSGlobalObject>(vm.heap)) JSGlobalObject(vm, structure, methodTable);
+    globalObject->finishCreation(vm);
+    return globalObject;
+}
```
Comment (philomates) [inline Source/JavaScriptCore/runtime/JSGlobalObject.cpp:2578]: this was the best way I could think of creating a new global object with a custom method table. We grab the method table from the original global object in https://github.com/WebKit/WebKit/pull/8/files#diff-44ebd0942590c3af4f75c8833b1b8f2dd9d649fd9e7dcc7a9e1410df5711c22fR33, which might be defined in `jsc.cpp` or elsewhere in webkit when run in the browser context.

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  let result = @evalInRealm(this, evalStr)
+  return @wrap(result);
+}
+
+function importValue(module, binding)
+{
+  let bindingStr = binding.toString();
+  let lookupBinding = (m) => {
+    if (bindingStr in m) {
+      return @wrap(m[bindingStr]);
```
Comment (philomates) [inline Source/JavaScriptCore/builtins/ShadowRealmPrototype.js:59]: is there a better way to do this prop check/lookup?

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
```
Comment (philomates) [inline Source/JavaScriptCore/builtins/ShadowRealmPrototype.js:25]: my (idiomatic) JS is horrible so in general be thorough with this file

File: Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp (hunk 4 above)
Comment (caiolima) [reply Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:84]: Those different global objects are really confusing my mind from time to time. I hope we have tests covering things there.

File: Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+    if (callFrame->argumentCount() != 2) {
+        throwTypeError(globalObject, scope, "Expected to be called with two arguments"_s);
+        RELEASE_AND_RETURN(scope, JSValue::encode(jsUndefined()));
+    }
+
+    JSValue thisValue = callFrame->argument(0);
+    ShadowRealmObject* thisRealm = jsDynamicCast<ShadowRealmObject*>(vm, thisValue);
+    if (UNLIKELY(!thisRealm))
+        return throwVMTypeError(globalObject, scope, "First arg should be a ShadowRealm");
```
Comment (caiolima) [inline Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:105]: s/arg/argument/

Comment (caiolima) [reply Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:105]: Since it’s internal, I was wondering if this could be an assert.

File: Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+    JSValue evalArg = callFrame->argument(1);
+    if (!evalArg.isString())
+        return throwVMTypeError(globalObject, scope, "Second argument must be a string");
```
Comment (caiolima) [inline Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:109]: ditto

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    @throwTypeError("value passing between realms must be callable or primitive");
+  }
+  return target;
+}
+
+function evaluate(evalStr)
+{
+  "use strict";
+
+  let result = @evalInRealm(this, evalStr)
```
Comment (caiolima) [inline Source/JavaScriptCore/builtins/ShadowRealmPrototype.js:50]: We need to validate here if both `this` and `evalStr` are valid inputs. otherwise, the errors from  `evalInRealm` won’t make much sense to  users and it’ll be exposing some internals that are not good to be exposed.

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  return @wrap(result);
+}
+
+function importValue(module, binding)
+{
```
Comment (caiolima) [inline Source/JavaScriptCore/builtins/ShadowRealmPrototype.js:55]: We also need to validate inputs here on JS level. This function can be called with anything, including `this` being overridden

File: Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+    JSGlobalObject* realmGlobalObject = thisRealm->globalObject();
+    NakedPtr<Exception> evaluationException;
+    SourceCode source = makeSource(sourceCode, callFrame->callerSourceOrigin(vm));
+    JSValue result = JSC::evaluate(realmGlobalObject, source, realmGlobalObject, evaluationException);
+    if (evaluationException) {
+        if (vm.isTerminationException(evaluationException.get()))
+            vm.setExecutionForbidden();
+
+        throwTypeError(globalObject, scope, "Error encountered during evaluation"_s);
```
Comment (caiolima) [inline Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:121]: I’m wondering if this exception will override the exception thrown by eval’ed code...

**Review by caiolima — CHANGES_REQUESTED** [review]
The patch is going in a good direction. I left some comments on things we need to change. I also think we need to add a feature flag for ShadowRealms and start writing stress tests, since Test262 is not run on Webkit’s CI for patches, only when they are landed.

File: Source/JavaScriptCore/runtime/ShadowRealmConstructor.cpp [hunk 13]
```diff
@@ -0,0 +1,49 @@
+#include "config.h"
+#include "ShadowRealmConstructor.h"
+
+#include "ShadowRealmObject.h"
+#include "JSCInlines.h"
+
+
+namespace JSC {
+
+const ClassInfo ShadowRealmConstructor::s_info = { "Function", &InternalFunction::s_info, nullptr, nullptr, CREATE_METHOD_TABLE(ShadowRealmConstructor) };
```
[the remark this answers is not in the document]
Comment (philomates) [reply Source/JavaScriptCore/runtime/ShadowRealmConstructor.cpp:10]: I wondered about that too but saw that other constructors use `Function`

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    delete wrapped['name'];
+    delete wrapped['length'];
+    return wrapped;
+  } else if (@isObject(target)) {
+    @throwTypeError("value passing between realms must be callable or primitive");
+  }
+  return target;
+}
+
+function evaluate(evalStr)
```
[the remark this answers is not in the document]
Comment (philomates) [reply Source/JavaScriptCore/builtins/ShadowRealmPrototype.js:46]: I didn't see a way to name this differently without changing the API of ShadowRealm itself

File: Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp (hunk 8 above)
Comment (philomates) [reply Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:105]: can you explain this a bit more; I see other `JSC_DEFINE_HOST_FUNCTION` that are throwing type errors instead of asserting

File: Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp (hunk 12 above)
Comment (philomates) [reply Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:121]: Like this?
```
> (new ShadowRealm()).evaluate("throw Error(\"hidden secret\")")
Exception: TypeError: Error encountered during evaluation
```

File: Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp (hunk 8 above)
Comment (philomates) [reply Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:105]: Is it that we can do an assert here because we should check this at the JS level?

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+ */
+
+@globalPrivate
+function wrap(target)
+{
+  "use strict";
+
+  if (@isCallable(target)) {
+    let wrapped = (...args) => {
+      const wrappedArgs = args.map(@wrap);
```
[the remark this answers is not in the document]
Comment (philomates) [reply Source/JavaScriptCore/builtins/ShadowRealmPrototype.js:33]: Hmm, I couldn't figure out how to use `@Map`. Any tips?

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+@globalPrivate
+function wrap(target)
```
[the remark this answers is not in the document]
Comment (philomates) [reply Source/JavaScriptCore/builtins/ShadowRealmPrototype.js:27]: So I tried to change this and the build kept breaking :/
I'm also having lots of issues updating this file, compiling, and getting results that indicate the changes weren't compiled; maybe that is related

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+  "use strict";
+
+  if (@isCallable(target)) {
+    let wrapped = (...args) => {
+      const wrappedArgs = args.map(@wrap);
+      const result = target.@apply(@undefined, wrappedArgs);
+      return @wrap(result);
+    };
+    delete wrapped['name'];
+    delete wrapped['length'];
```
[the remark this answers is not in the document]
Comment (philomates) [reply Source/JavaScriptCore/builtins/ShadowRealmPrototype.js:38]: I was going off of this:
```
> var x = () => { 1 }
undefined
> Object.getOwnPropertyNames(x)
length,name
```

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+
+  let result = @evalInRealm(this, evalStr)
+  return @wrap(result);
+}
+
+function importValue(module, binding)
+{
+  let bindingStr = binding.toString();
+  let lookupBinding = (m) => {
+    if (bindingStr in m) {
```
[the remark this answers is not in the document]
Comment (philomates) [reply Source/JavaScriptCore/builtins/ShadowRealmPrototype.js]: do modules have a prototype chain? 

Like if I run 
```
 var r = new ShadowRealm();
 r.evaluate("Object.prototype.whatever = 1");
 await r.importValue("./fruit.js", "whatever");
```
where `fruit.js` doesn't have `whatever` defined/exported, it fails with
```
Exception: TypeError: %ShadowRealm%.importValue requires |exportName| to exist in the |specifier|
```

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+
+function evaluate(sourceText)
+{
+  "use strict";
+
+  if (!@isShadowRealm(this))
+      @throwTypeError("`%ShadowRealm%.evaluate requires that |this| be a ShadowRealm instance");
+
+  if (typeof sourceText !== 'string')
+      @throwTypeError("`%ShadowRealm%.evaluate requires that the |sourceText| argument be a string");
```
Comment (philomates) [inline Source/JavaScriptCore/builtins/ShadowRealmPrototype.js:54]: is this robust enough or should we also do `|| sourceText instanceof String` to get the case where a `new String(123)` arg is passed in?

File: Source/JavaScriptCore/runtime/JSGlobalObject.cpp [hunk 20]
```diff
@@ -1242,6 +1253,9 @@ capitalName ## Constructor* lowerName ## Constructor = featureFlag ? capitalName
         TemporalObject* temporal = TemporalObject::create(vm, TemporalObject::createStructure(vm, this));
         putDirectWithoutTransition(vm, vm.propertyNames->Temporal, temporal, static_cast<unsigned>(PropertyAttribute::DontEnum));
     }
+    if (Options::useShadowRealm()) {
+        putDirectWithoutTransition(vm, vm.propertyNames->ShadowRealm, shadowRealmConstructor, static_cast<unsigned>(PropertyAttribute::DontEnum));
+    }
```
Comment (philomates) [inline Source/JavaScriptCore/runtime/JSGlobalObject.cpp:1258]: should I put more in this feature flag conditional?

File: Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+
+JSC_DEFINE_HOST_FUNCTION(importInRealm, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    if (callFrame->argumentCount() != 2) {
+        throwTypeError(globalObject, scope, "Expected to be called with two arguments"_s);
+        RELEASE_AND_RETURN(scope, JSValue::encode(jsUndefined()));
+    }
```
Comment (philomates) [inline Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:70]: since `importInRealm` is called by a builtin, we can probably remove this check, right?

File: Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp (hunk 3 above)
Comment (caiolima) [reply Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp]: It’s really a question I don’t have the answer...I didn’t even know we had both.

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js (hunk 6 above)
Comment (caiolima) [reply Source/JavaScriptCore/builtins/ShadowRealmPrototype.js]: Not in this computed case. But such access will be optimized out on DFG and FTL if possible.

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (caiolima) [reply Source/JavaScriptCore/builtins/ShadowRealmPrototype.js:27]: Hum...I remember having troubles like that before. I used to force a clean build IIRC.

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (caiolima) [reply Source/JavaScriptCore/builtins/ShadowRealmPrototype.js:33]: I’m sorry, I meant `@map`. If this is not around, we will probably need to include `mapPrivateName` on Array prototypes to make it available here.

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (caiolima) [reply Source/JavaScriptCore/builtins/ShadowRealmPrototype.js]: I don’t think they have. I remember people saying em that modules are different than common objects. If this is the case, I think we might be fine.

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (caiolima) [reply Source/JavaScriptCore/builtins/ShadowRealmPrototype.js:38]: Oh right, `Object.getOwnPropertyNames` is only going to show enumerable properties IIRC. We probably need to take a look at spec to be sure there.

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (caiolima) [reply Source/JavaScriptCore/builtins/ShadowRealmPrototype.js]: Oh, I think this is fine to keep it evaluate. My bad, because I thought it was a global operation.

File: Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp (hunk 8 above)
Comment (caiolima) [reply Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp]: That’s what I was thinking about. How those errors show up for users when they call `ShadowRealm.prototype.evaluate(1, Symbol(“foo”));`, fox example?

File: Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp (hunk 12 above)
Comment (caiolima) [reply Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp]: Yes, just like that. I think you shared this internally, and spec says we should override it, right?

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js (hunk 19 above)
Comment (caiolima) [reply Source/JavaScriptCore/builtins/ShadowRealmPrototype.js:54]: Spec says `If Type(sourceText) is not String, throw a TypeError exception.`, so I think for String Objects (like `new String(123)`) we should throw, because they are considered Objects. Might be worth it validate this with spec experts!

File: Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp (hunk 21 above)
Comment (caiolima) [reply Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:70]: I think so. Placing an assert might make bugs there more obvious, BTW

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (philomates) [reply Source/JavaScriptCore/builtins/ShadowRealmPrototype.js]: `@map` doesn't exist and I couldn't figure out how `mapPrivate..` might apply.
what about this: https://github.com/philomates/WebKit/pull/1/commits/79b31eda9fec2b54db6857fc293485d3c9f6dea3 ?

File: Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp (hunk 8 above)
Comment (philomates) [reply Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp]: I switched it to an ASSERT and did the check on the JS side and it looks like this now, which seems correctl:
```
ShadowRealm.prototype.evaluate(1, Symbol("foo"));
Exception: TypeError: `%ShadowRealm%.evaluate requires that |this| be a ShadowRealm instance
```

File: Source/JavaScriptCore/builtins/ShadowRealmPrototype.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (philomates) [reply Source/JavaScriptCore/builtins/ShadowRealmPrototype.js:48]: I looked at `FunctionPrototype` and `JSFunction` and didn't see anything else defined, aside from `name` and `length`

**Review by caiolima — COMMENTED** [review]
Patch LGTM. I think it’s ready to be shared on Bugzilla and ask for some review already.

File: JSTests/stress/shadow-realm-evaluate.js [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+{
+    let realm = new ShadowRealm();
+    shouldThrow(
+      () => { realm.evaluate(".."); },
+      SyntaxError,
+      (err) => {
+          shouldBe($.globalObjectFor(err), globalThis);
+          shouldBe(String(err), `SyntaxError: Unexpected token '.'`);
+      });
+}
```
[the remark this answers is not in the document]
Comment (philomates) [reply JSTests/stress/shadow-realm-evaluate.js:61]: I'll ask Leo about it

[the remark this answers is not in the document]
Comment (leobalter) [reply JSTests/stress/shadow-realm-evaluate.js:60]: We have a [very shy assertion](https://github.com/tc39/test262/blob/main/test/built-ins/ShadowRealm/prototype/evaluate/errors-from-the-other-realm-is-wrapped-into-a-typeerror.js#L18) picking the SyntaxError. 

The assert.throws in Test262 verifies the identity of the expected error constructor being Realm related.

I believe this is a good opportunity to add an [assertion for assert.throws](https://github.com/tc39/test262/tree/main/test/harness) making sure the `err.constructor === expectedErrCtor` in order to avoid identity discontinuity issues.

[the remark this answers is not in the document]
Comment (leobalter) [reply JSTests/stress/shadow-realm-evaluate.js:60]: For the records, we now make sure that assert.throws won't give false positive for cross realms error constructors. https://github.com/tc39/test262/pull/3208

[the remark this answers is not in the document]
Comment (leobalter) [reply JSTests/stress/shadow-realm-evaluate.js:60]: which means that in Test262 [tests like this](https://github.com/tc39/test262/blob/main/test/built-ins/ShadowRealm/prototype/evaluate/errors-from-the-other-realm-is-wrapped-into-a-typeerror.js#L18-L21) are good enough for their purposes of verifying the proper identity.

[the remark this answers is not in the document]
Comment (philomates) [reply JSTests/stress/shadow-realm-evaluate.js:60]: Nice!
Come to think of it, I was getting test262 failures last week essentially saying "expected a `SyntaxError` but got a `SyntaxError`". At the time I was super confused but now I'm realizing that it was the bug in my implementation where errors created in the child realm using one global object were being passed back unwrapped and being compared to errors in the parent realm, which has a different global object. Being new to JS I didn't realize that this test failure could imply this issues but I'm now seeing the connection after writing the explicit `$.globalObjectFor(err)` checks. 
Long story short, test262 seemed to be correctly failing when errors escaped their realms unmodified.

**Comment by philomates** [comment]
moved to https://bugs.webkit.org/show_bug.cgi?id=230602
